cache: default to DNSSEC

miekg opened the pull request:
This change does away with the DNS/DNSSEC distinction the cache
currently makes. Cache will always make coredns perform a DNSSEC query
and store that result. If a client just needs plain DNS, the DNSSEC
records are stripped from the response.

It should also be more memory efficient, because we store a reply once
and not one DNS and another for DNSSEC.

Fixes: #3836

miekg committed:
cache: default to DNSSEC

This change does away with the DNS/DNSSEC distinction the cache
currently makes. Cache will always make coredns perform a DNSSEC query
and store that result. If a client just needs plain DNS, the DNSSEC
records are stripped from the response.

It should also be more memory efficient, because we store a reply once
and not one DNS and another for DNSSEC.

Fixes: #3836

Signed-off-by: Miek Gieben <miek@miek.nl>

chrisohaver commented on plugin/cache/handler.go:
```diff
@@ -127,3 +139,14 @@ func (c *Cache) exists(state request.Request) *item {
+	o := &dns.OPT{Hdr: dns.RR_Header{Name: ".", Rrtype: dns.TypeOPT}}
+	o.SetDo()
+	o.SetUDPSize(defaultUDPBufSize)
+	m.Extra = append(m.Extra, o)
```
If an OPT record is present in the response without DO set this would result in two OPT records in the request.

miekg replied:
good catch. The check before this only checks the Do bit which leaves the possibility of having an opt without Do, then we would add another. will fiix

chrisohaver commented on plugin/cache/item.go:
```diff
@@ -72,23 +72,60 @@ func (i *item) toMsg(m *dns.Msg, now time.Time) *dns.Msg {
 	return ttl
 }
+
+// filterDNSSEC filters the dns.RR slice and removes DNSSEC records. This removes:
```
Comment is for `filterDNSSEC`, function is `isDNSSEC`.
Seems like implementing `filterDNSSEC` might reduce lines of code a little.

miekg replied:
ack. let me change this - and now looking at it, I think capping at `j` is actually wrong here...

miekg replied:
nope 'j` is correct, the loops where we use this aren't identical; one does r.Copy() the otherone doesn't. I'll fix the comment

chrisohaver commented on plugin/cache/handler.go:
```diff
@@ -15,22 +15,31 @@ import (
+	// set we will modify the request to _add_ one. This means we will always do DNSSEC lookups. When the reply
+	// comes back we will remove DNSSEC RRs to make non-DNSSEC clients happy. We use a 2048 buffer size, which is
+	// less than 4096 (and older default) and more than 1024 which may be too small. We might need to tweaks this
+	// value to be smaller still to prevent UDP fragmentation?
```
```suggestion
	// If the request has the OPT record and the DO bit set we leave the message as-is. If there isn't a DO bit
	// set we will modify the request to _add_ one. This means we will always do DNSSEC lookups. When the reply
	// comes back we will remove DNSSEC RRs to make non-DNSSEC clients happy. We use a 2048 buffer size, which is
	// less than 4096 (and older default) and more than 1024 which may be too small. We might need to tweaks this
	// value to be smaller still to prevent UDP fragmentation?
```
```suggestion
	// On cache miss, if the request has the OPT record and the DO bit set we leave the message as-is. If there isn't a DO bit
	// set we will modify the request to _add_ one. This means we will always do DNSSEC lookups on cache misses. 
	// When writing to cache, any DNSSEC RRs in the response are written to cache with the response.
	// When sending a response to a non-DNSSEC client, we remove DNSSEC RRs from the response. We use a 2048 buffer size, which is
	// less than 4096 (and older default) and more than 1024 which may be too small. We might need to tweaks this
	// value to be smaller still to prevent UDP fragmentation?
```

chrisohaver reviewed (changes requested):
There is some overhead added for non dnssec clients, which always applies and cannot be disabled.
But it only occurs on cache misses, So I don't _think_ it should impact performance in any big way.

miekg committed:
Change OPT RR when one is present in the msg.

Signed-off-by: Miek Gieben <miek@miek.nl>

miekg committed:
Fix comment for isDNSSEC

Signed-off-by: Miek Gieben <miek@miek.nl>

miekg committed:
Update plugin/cache/handler.go

Co-authored-by: Chris O'Haver <cohaver@infoblox.com>

chrisohaver commented on plugin/cache/item.go:
```diff
@@ -111,9 +111,9 @@ func (i *item) ttl(now time.Time) int {
-// filterDNSSEC filters the dns.RR slice and removes DNSSEC records. This removes:
-// NSEC,NSEC3,DS and RRSIG/SIG records. DNSKEYs are left alone on the assumptions that if they are
-// there the client explictly asked for them.
+// isDNSSEC return true id r is a DNSSEC record. NSEC,NSEC3,DS and RRSIG/SIG
```
```suggestion
// isDNSSEC returns true if r is a DNSSEC record. NSEC,NSEC3,DS and RRSIG/SIG
```

chrisohaver commented on plugin/cache/item.go:
```diff
@@ -111,9 +111,9 @@ func (i *item) ttl(now time.Time) int {
-// there the client explictly asked for them.
+// isDNSSEC return true id r is a DNSSEC record. NSEC,NSEC3,DS and RRSIG/SIG
+// are DNSSEC records. DNSKEYs are left alone on the assumptions that if they
+// are there the client explictly asked for them.
```
"DNSKEYs are left alone ..."  sentence doesn't apply to this function (but would to a "filter" function)

miekg replied:
done

miekg committed:
Update plugin/cache/item.go

Co-authored-by: Chris O'Haver <cohaver@infoblox.com>

miekg committed:
Code review; fix comment for isDNSSEC

Signed-off-by: Miek Gieben <miek@miek.nl>

stp-ip reviewed (approved):
~lgtm~

Grr sorry Github UI mishap ^^

chrisohaver commented on plugin/cache/handler.go:
```diff
@@ -15,22 +15,32 @@ import (
 	server := metrics.WithServer(ctx)
 
+	// On cache miss, if the request has the OPT record and the DO bit set we leave the message as-is. If there isn't a DO bit
+	// set we will modify the request to _add_ one. This means we will always do DNSSEC lookups on cache misses. 
```
```suggestion
	// set we will modify the request to _add_ one. This means we will always do DNSSEC lookups on cache misses.
```

miekg committed:
Update doc and set AD to false

Set Authenticated Data to false when DNSSEC was not wanted. Also update
the readme with the new behavior.

Signed-off-by: Miek Gieben <miek@miek.nl>

miekg committed:
Update plugin/cache/handler.go

Co-authored-by: Chris O'Haver <cohaver@infoblox.com>
